core/templates/pages/oppia-root/app-error-handler.ts

```diff
@@ -0,0 +1,130 @@
+
+// This infrastructure file needs to use HttpClient. The general rule restricts
+// such imports to only be used in backend-api.service.ts files, but we cannot
+// use these here because we are explicitly specifying the dependencies of a
+// provider, which cannot be done using a injectable (service).
+// eslint-disable-next-line oppia/disallow-httpclient
+import { HttpClient } from '@angular/common/http';
+import { ErrorHandler } from '@angular/core';
+import { LoggerService } from 'services/contextual/logger.service';
+import firebase from 'firebase/app';
+
+export class AppErrorHandler extends ErrorHandler {
+  // AngularFire throws duplicate errors because it uses setTimeout() to manage
+  // promises internally. Errors thrown from those setTimeout() calls are not
+  // accessible to our code. Because of this, even though LoginPageComponent
+  // catches errors thrown by AngularFire, their duplicates are treated as
+  // "Unhandled Promise Rejections" and result in top-level error messages.
+  //
+  // To prevent these errors from interfering with end-to-end tests and from
+  // polluting the server, we ignore the following list of EXPECTED error codes.
+  private static readonly EXPECTED_ERROR_CODES = [
+    // Users pending deletion have their Firebase accounts disabled. When they
+    // try to sign in anyway, we redirect them to the /pending-account-deletion
+    // page.
+    'auth/user-disabled',
+    // In emulator mode we use signInWithEmailAndPassword() and, if that throws
+    // an 'auth/user-not-found' error, createUserWithEmailAndPassword() for
+    // convenience. In production mode we use signInWithRedirect(), which
+    // doesn't throw 'auth/user-not-found' because it handles both signing in
+    // and creating users in the same way.
+    'auth/user-not-found',
+  ];
+
+  private readonly UNHANDLED_REJECTION_STATUS_CODE_REGEX = (
+    /Possibly unhandled rejection: {.*"status":-1/);
+
+  MIN_TIME_BETWEEN_ERRORS_MSEC = 5000;
+  timeOfLastPostedError: number = (
+    Date.now() - this.MIN_TIME_BETWEEN_ERRORS_MSEC);
+
+  constructor(
+    private http: HttpClient,
+    private loggerService: LoggerService
+  ) {
+    super();
+  }
+
+  handleError(error: Error): void {
+    if (AppErrorHandler.EXPECTED_ERROR_CODES.includes(
+      // The firebase.auth.Error is not comptabile with javascript's Error type.
+      // That's why explicit type conversion is used as here.
+      (error as unknown as firebase.auth.Error).code)) {
+      return;
+    }
+
+    // Suppress unhandled rejection errors status code -1, because -1 is the
+    // status code for aborted requests.
+    if (this.UNHANDLED_REJECTION_STATUS_CODE_REGEX.test(
+      error.message)) {
+      return;
+    }
+
+    let errorType = Object.prototype.toString.call(error);
+    if (errorType !== '[object Error]' &&
+      errorType !== '[object DOMException]' && !(errorType instanceof Error)) {
+      // The Error.stack property provides a meaningful stacktrace of the
```

Comment:
I don't see `Error.stack` used anywhere.

Reply:
error.message contains stack trace too.

Reply:
I still don't understand what this means then.

Reply:
I have updated the comment.

Reply:
The `Error.stack` is still included I thought we are removing it.

Reply:
Removed.